packages/runtime/src/renderer.ts

```diff
@@ -172,6 +173,48 @@ export function getTarget(potentialTarget: object): object {
   return potentialTarget;
 }
 
+export function getRefTarget(refHost: ComponentHost<INode>, refTargetName: string): object {
+  if (refTargetName === 'element') {
+    return refHost;
+  }
+  const $auRefs = refHost.$au;
+  if ($auRefs === void 0) {
+    // todo: code error code, this message is from v1
+    throw new Error(`No Aurelia APIs are defined for the element: "${(refHost as { tagName: string }).tagName}".`);
+  }
+  switch (refTargetName) {
+    case 'controller':
+      // this means it supports returning undefined
+      return (refHost as CustomElementHost<INode>).$controller as IController;
+    case 'view':
+      // todo: returns node sequences for fun?
+      throw new Error('Not supported API');
+    case 'view-model':
+        // this means it supports returning undefined
+        return ((refHost as CustomElementHost<INode>).$controller as IController).viewModel!;
+    default:
+      const refTargetController = $auRefs[refTargetName];
+      if (refTargetController === void 0) {
+        throw new Error(`Attempted to reference "${refTargetName}", but it was not found amongst the target's API.`);
+      }
+      return refTargetController.viewModel!;
+  }
+}
+
+function setControllerReference<T = INode>(
+  controller: Controller<T>,
+  host: ComponentHost<T>,
+  referenceName: string
+): void {
+  let $auRefs = host.$au;
+  if ($auRefs === void 0) {
+    $auRefs = host.$au = new ControllersLookup() as Record<string, IController<T>>;
+  }
+  $auRefs[referenceName] = controller;
+}
+
+class ControllersLookup {}
```

Comment:
```suggestion
class ControllersLookup<T extends INode> {
  public [key: string]: IController<T>;
}
```

Then you don't need the `as Record<string, IController<T>>` cast